File: cpp/questions/minimumCost3.cpp
#include <vector>
#include <set>

using namespace std;

class Solution {
public:
    long long minimumCost(vector<int>& nums, int k, int dist) {
        std::set<std::pair<int, int>> area;
        for (int p=1; p<=1+dist; p++) {
            area.insert({nums[p], p});
        }
        auto it = area.begin();
        long long curs = nums[0];
        for (int p=0; p<k-1; p++) {
            curs += it->first;
            it++;
        }
        it--;
        long long ans = curs;
        for (int p=2+dist; p<nums.size(); p++) {
            std::pair<int, int> ie = {nums[p], p};
            std::pair<int, int> de = {nums[p-dist-1], p-dist-1};
            area.insert(ie);
            if (ie.first < it->first) {
                curs += ie.first - (it--)->first;
            }
            if (de.first <= it->first) {
                curs += (++it)->first - de.first;
            }
            area.erase(de);
            ans = std::min(ans, curs);
        }
        return ans;
    }
};

File: python/questions/minimumCost3.py
from typing import List
from sortedcontainers import SortedList

class Solution:
    def minimumCost(self, nums: List[int], k: int, dist: int) -> int:
        area = SortedList()
        curs = nums[0]
        for p in range(1, 2+dist):
            area.add(nums[p])
        curs += sum(area[:k-1]) 
        ans = curs
        for p in range(2+dist, len(nums)):
            ie = nums[p]
            de = nums[p-dist-1]
            area.add(nums[p])
            if ie <= area[k-2]:
                curs += ie - area[k-1]
            if de <= area[k-2]:
                curs += area[k-1] - de
            area.remove(de)
            ans = min(ans, curs)
        return ans
